util/pingpong.c

```diff
@@ -0,0 +1,2229 @@
+#endif
+
+enum precision {
+	NANO = 1,
+	MICRO = 1000,
+	MILLI = 1000000,
+};
+
+enum {
+	PP_OPT_ACTIVE		= 1 << 0,
+	PP_OPT_ITER		= 1 << 1,
+	PP_OPT_SIZE		= 1 << 2,
+	PP_OPT_VERIFY_DATA	= 1 << 3,
+};
+
+struct pp_opts {
+	char *src_port;
+	char *dst_port;
+	char *src_addr;
+	char *dst_addr;
+	int iterations;
+	int transfer_size;
+	int sizes_enabled;
+	int options;
+	int argc;
+	char **argv;
+};
+
+#define PP_SIZE_MAX_POWER_TWO 22
+#define PP_MAX_DATA_MSG (1 << (PP_SIZE_MAX_POWER_TWO + 1)) + (1 << PP_SIZE_MAX_POWER_TWO);
+
+#define PP_STR_LEN 32
+#define PP_MAX_CTRL_MSG 64
+#define PP_CTRL_BUF_LEN 64
+#define PP_MR_KEY 0xC0DE
+
+#define INTEG_SEED 7
+#define PP_ENABLE_ALL (~0)
+#define PP_DEFAULT_SIZE (1 << 0)
+
+#define PP_MSG_CHECK_PORT_OK "port ok"
+#define PP_MSG_LEN_PORT 5
+#define PP_MSG_CHECK_CNT_OK "cnt ok"
+#define PP_MSG_LEN_CNT 10
+#define PP_MSG_SYNC_Q "q"
+#define PP_MSG_SYNC_A "a"
+
+#define PP_PRINTERR(call, retv)						\
+	do {								\
+		fprintf(stderr, "%s(): %s:%-4d, ret=%d (%s)\n", call,	\
+			__FILE__, __LINE__, (int) retv,			\
+			fi_strerror((int) -retv));			\
+	} while (0)
+
+#define PP_ERR(fmt, ...)						\
+	do {								\
+		fprintf(stderr, "[%s] %s:%-4d: " fmt "\n", "error",	\
+			__FILE__, __LINE__, ##__VA_ARGS__);		\
+	} while (0)							\
+
+int PP_ACTIVATE_DEBUG = 0;
+
+#define PP_DEBUG(fmt, ...)						\
+	if (PP_ACTIVATE_DEBUG) {					\
+		fprintf(stderr, "[%s] %s:%-4d: " fmt, "debug",		\
+			__FILE__, __LINE__, ##__VA_ARGS__);		\
+	}								\
+
+#define PP_CLOSE_FID(fd)						\
+	do {								\
+		int ret;						\
+		if ((fd)) {						\
+			ret = fi_close(&(fd)->fid);			\
+			if (ret)					\
+				PP_ERR("fi_close (%d) fid %d",		\
+					ret, (int) (fd)->fid.fclass);	\
+			fd = NULL;					\
+		}							\
+	} while (0)
+
+#define MAX(a,b) (((a)>(b))?(a):(b))
+
+struct ct_pingpong {
+	struct fi_info *fi_pep, *fi, *hints;
+	struct fid_fabric *fabric;
+	struct fid_domain *domain;
+	struct fid_pep *pep;
+	struct fid_ep *ep;
+	struct fid_cq *txcq, *rxcq;
+	struct fid_mr *mr;
+	struct fid_av *av;
+	struct fid_eq *eq;
+
+	struct fid_mr no_mr;
+	struct fi_context tx_ctx, rx_ctx;
+	struct fi_context *ctx_arr;
+	uint64_t remote_cq_data;
+
+	uint64_t tx_seq, rx_seq, tx_cq_cntr, rx_cq_cntr;
+	int pp_parent_proc;
+	pid_t pp_child_pid;
+
+	fi_addr_t remote_fi_addr;
+	void *buf, *tx_buf, *rx_buf;
+	size_t buf_size, tx_size, rx_size;
+	int rx_fd, tx_fd;
+	int data_default_port;
+	char data_port[8];
+
+	char test_name[50];
+	int timeout;
+	struct timespec start, end;
+
+	struct fi_av_attr av_attr;
+	struct fi_eq_attr eq_attr;
+	struct fi_cq_attr cq_attr;
+	struct pp_opts opts;
+
+	long cnt_ack_msg;
+
+	int ctrl_port;
+	int ctrl_listenfd;
+	int ctrl_connfd;
+	struct sockaddr_in ctrl_addr;
+	char ctrl_buf[PP_CTRL_BUF_LEN + 1];
+};
+
+static const char integ_alphabet[] = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+static const int integ_alphabet_length = (sizeof(integ_alphabet)/sizeof(*integ_alphabet)) - 1;
```

Comment:
Because `integ_alphabet` has the null byte at the end, and we do not want to use that for the verification mode.
To be thorough :
```
sizeof(integ_alphabet)  == 10 + 26 + 26 + 1 == 63
sizeof(*integ_alphabet)) == 1
(sizeof(integ_alphabet)/sizeof(*integ_alphabet)) == 63
(sizeof(integ_alphabet)/sizeof(*integ_alphabet)) - 1 == 62
```
So `62` is the length of the part of the array with the printable characters that we want to use.